In [1]:
import pandas as pd
import numpy as np
import xgboost as xgb
import yfinance as yf
from datetime import datetime, timedelta
import requests
from io import StringIO
import matplotlib.pyplot as plt

In [2]:
# Load the CSV file into a DataFrame
df = pd.read_csv('mega_large_tickers.csv')

# Filter stocks with ticker symbols of 4 letters or less
df = df[df['Symbol'].str.len() <= 4]

# Rename the 'Symbol' column to 'ticker' (lowercase)
df.rename(columns={'Symbol': 'ticker'}, inplace=True)

# Save the filtered DataFrame to a new CSV file
df.to_csv('filtered_tickers.csv', index=False)

# Optionally, display the first few rows of the filtered DataFrame
print(df.head())

  ticker
0      A
1   AAPL
2   ABBV
3   ABEV
4   ABNB


In [3]:
import yfinance as yf
import pandas as pd

# Function to download historical data for a single ticker with start and end dates
def download_data(ticker, start, end):
    data = yf.download(ticker, start=start, end=end)
    if 'dst_error_hours' in data.columns:
        # Correct the timedelta handling
        data.index += pd.to_timedelta(data['dst_error_hours'], unit='h')
        data.drop(columns='dst_error_hours', inplace=True)
    data['ticker'] = ticker
    return data

# Define the start and end dates for the data download
start_date = '2019-07-20'  # Example start date
end_date = '2024-07-20'    # Example end date

# Assuming your dataframe "df" has a column named 'ticker' with the list of tickers
tickers = df['ticker'].tolist()

# Download data for all tickers and concatenate into a single DataFrame
data_frames = [download_data(ticker, start_date, end_date) for ticker in tickers]
all_data = pd.concat(data_frames)

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%*******

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%*******

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%*******

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%*******

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%*******

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%*******

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%*******

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


In [4]:
all_data.head()

Open       High        Low      Close  Adj Close   Volume  \
Date                                                                         
2019-07-22  68.910004  69.470001  68.459999  69.089996  66.713165  3103300   
2019-07-23  69.400002  70.169998  69.160004  69.889999  67.485634  3304200   
2019-07-24  69.250000  69.580002  68.620003  69.580002  67.186302  3190400   
2019-07-25  69.750000  70.010002  69.550003  69.849998  67.447021  1904700   
2019-07-26  70.129997  70.190002  69.349998  69.500000  67.109070  2392600   

           ticker  
Date               
2019-07-22      A  
2019-07-23      A  
2019-07-24      A  
2019-07-25      A  
2019-07-26      A

In [5]:
# Save the fixed DataFrame to a new CSV file, preserving the multi-index
all_data.to_csv('temporary_data.csv', index=True)

In [6]:
# Read the CSV file, with only Date as the index and ticker as a regular column
all_data = pd.read_csv('temporary_data.csv', index_col='Date', parse_dates=['Date'])

In [7]:
all_data.head()

Open       High        Low      Close  Adj Close   Volume  \
Date                                                                         
2019-07-22  68.910004  69.470001  68.459999  69.089996  66.713165  3103300   
2019-07-23  69.400002  70.169998  69.160004  69.889999  67.485634  3304200   
2019-07-24  69.250000  69.580002  68.620003  69.580002  67.186302  3190400   
2019-07-25  69.750000  70.010002  69.550003  69.849998  67.447021  1904700   
2019-07-26  70.129997  70.190002  69.349998  69.500000  67.109070  2392600   

           ticker  
Date               
2019-07-22      A  
2019-07-23      A  
2019-07-24      A  
2019-07-25      A  
2019-07-26      A

In [8]:
# Reset the index to include only 'ticker' for filling NaNs correctly
all_data.reset_index(inplace=True)
all_data.set_index('ticker', inplace=True)

# Handle NaNs: forward fill, then backward fill within each ticker group
all_data = all_data.groupby('ticker').ffill()
all_data = all_data.groupby('ticker').bfill()

In [9]:
all_data.head()

Date       Open       High        Low      Close  Adj Close  \
ticker                                                                     
A      2019-07-22  68.910004  69.470001  68.459999  69.089996  66.713165   
A      2019-07-23  69.400002  70.169998  69.160004  69.889999  67.485634   
A      2019-07-24  69.250000  69.580002  68.620003  69.580002  67.186302   
A      2019-07-25  69.750000  70.010002  69.550003  69.849998  67.447021   
A      2019-07-26  70.129997  70.190002  69.349998  69.500000  67.109070   

         Volume  
ticker           
A       3103300  
A       3304200  
A       3190400  
A       1904700  
A       2392600

In [10]:
import pandas as pd

# Ensure 'Date' is in datetime format
all_data['Date'] = pd.to_datetime(all_data['Date'])

In [11]:
all_data.tail()

Date        Open        High         Low       Close   Adj Close  \
ticker                                                                          
ZTS    2024-07-15  178.580002  180.850006  177.820007  177.979996  177.557648   
ZTS    2024-07-16  178.100006  183.220001  177.779999  182.910004  182.475967   
ZTS    2024-07-17  182.000000  184.839996  180.559998  182.050003  181.617996   
ZTS    2024-07-18  180.559998  184.240005  179.169998  179.229996  179.229996   
ZTS    2024-07-19  180.039993  180.509995  176.169998  178.580002  178.580002   

         Volume  
ticker           
ZTS     2216500  
ZTS     1939900  
ZTS     2190800  
ZTS     2171100  
ZTS     2131000

In [12]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1012371 entries, A to ZTS
Data columns (total 7 columns):
 #   Column     Non-Null Count    Dtype         
---  ------     --------------    -----         
 0   Date       1012371 non-null  datetime64[ns]
 1   Open       1012371 non-null  float64       
 2   High       1012371 non-null  float64       
 3   Low        1012371 non-null  float64       
 4   Close      1012371 non-null  float64       
 5   Adj Close  1012371 non-null  float64       
 6   Volume     1012371 non-null  int64         
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 61.8+ MB


In [13]:
# Convert the Date column to a DateTime index and set the multi-index
all_data.reset_index(inplace=True)
all_data['Date'] = pd.to_datetime(all_data['Date'])
all_data.set_index(['Date', 'ticker'], inplace=True)

# Display the resulting DataFrame to verify the multi-index structure
all_data.head()

Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-22 A       68.910004  69.470001  68.459999  69.089996  66.713165   
2019-07-23 A       69.400002  70.169998  69.160004  69.889999  67.485634   
2019-07-24 A       69.250000  69.580002  68.620003  69.580002  67.186302   
2019-07-25 A       69.750000  70.010002  69.550003  69.849998  67.447021   
2019-07-26 A       70.129997  70.190002  69.349998  69.500000  67.109070   

                    Volume  
Date       ticker           
2019-07-22 A       3103300  
2019-07-23 A       3304200  
2019-07-24 A       3190400  
2019-07-25 A       1904700  
2019-07-26 A       2392600

In [14]:
# Calculate daily returns
all_data['Daily_Return'] = all_data.groupby('ticker')['Adj Close'].pct_change()

# Calculate weekly returns
all_data['3_Day_Return'] = all_data.groupby('ticker')['Adj Close'].pct_change(3)

# Create a binary variable for Weekly_Return
all_data['3Day_Ret_Class'] = all_data['3_Day_Return'].apply(lambda x: 1 if x > 0 else 0)

# Calculate weekly returns
all_data['Weekly_Return'] = all_data.groupby('ticker')['Adj Close'].pct_change(5)

# Create a binary variable for Weekly_Return
all_data['Weekly_Ret_Class'] = all_data['Weekly_Return'].apply(lambda x: 1 if x > 0 else 0)

# Display the resulting DataFrame to verify the calculations
all_data.head()

Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-22 A       68.910004  69.470001  68.459999  69.089996  66.713165   
2019-07-23 A       69.400002  70.169998  69.160004  69.889999  67.485634   
2019-07-24 A       69.250000  69.580002  68.620003  69.580002  67.186302   
2019-07-25 A       69.750000  70.010002  69.550003  69.849998  67.447021   
2019-07-26 A       70.129997  70.190002  69.349998  69.500000  67.109070   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2019-07-22 A       3103300           NaN           NaN               0   
2019-07-23 A       3304200      0.011579           NaN               0   
2019-07-24 A       3190400     -0.004435           NaN               0   
2019-07-25 A       1904700      0.003881       0.01100               1   
2019-07-26 A       2392600     -0.005011      -0.00558               0   

                   Weekly_Return  Weekly_Ret_Class  
Date       ticker                                   
2019-07-22 A                 NaN                 0  
2019-07-23 A                 NaN                 0  
2019-07-24 A                 NaN                 0  
2019-07-25 A                 NaN                 0  
2019-07-26 A                 NaN                 0

In [15]:
# Calculate the range of the weekly returns
min_weekly_return = all_data['Weekly_Return'].min() * 100
max_weekly_return = all_data['Weekly_Return'].max() * 100
range_weekly_return = max_weekly_return - min_weekly_return

# Display the range of weekly returns
print(f"Minimum Weekly Return: {min_weekly_return:.2f}%")
print(f"Maximum Weekly Return: {max_weekly_return:.2f}%")
print(f"Range of Weekly Returns: {range_weekly_return:.2f}%")

Minimum Weekly Return: -90.00%
Maximum Weekly Return: 900.00%
Range of Weekly Returns: 990.00%


In [16]:
# Define thresholds for identifying outliers
min_threshold = -.5  # -50%
max_threshold = .5  # 50%

# Filter out the outliers
all_data = all_data[(all_data['Weekly_Return'] > min_threshold) & (all_data['Weekly_Return'] < max_threshold)]

# Save the filtered DataFrame to a new CSV file, preserving the multi-index
all_data.to_csv('filtered_data.csv', index=True)

# Display the resulting DataFrame to verify the filtering
all_data.head()


Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
2019-07-30 A       69.250000  70.180000  68.099998  70.150002  67.736710   
2019-07-31 A       70.190002  70.660004  68.919998  69.410004  67.022156   
2019-08-01 A       69.629997  70.680000  68.389999  69.209999  66.829041   
2019-08-02 A       69.120003  69.129997  67.559998  68.080002  65.737915   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2019-07-29 A       2750000      0.012518      0.011354               1   
2019-07-30 A       2871800     -0.003126      0.004295               1   
2019-07-31 A       3846100     -0.010549     -0.001295               0   
2019-08-01 A       2494100     -0.002881     -0.016484               0   
2019-08-02 A       2803500     -0.016327     -0.029508               0   

                   Weekly_Return  Weekly_Ret_Class  
Date       ticker                                   
2019-07-29 A            0.018527                 1  
2019-07-30 A            0.003720                 1  
2019-07-31 A           -0.002443                 0  
2019-08-01 A           -0.009162                 0  
2019-08-02 A           -0.020432                 0

In [17]:
# Find the rows with the minimum and maximum weekly returns
min_weekly_return_row = all_data[all_data['Weekly_Return'] == all_data['Weekly_Return'].min()]
max_weekly_return_row = all_data[all_data['Weekly_Return'] == all_data['Weekly_Return'].max()]

# Display the tickers and dates with the extreme weekly returns
print("Ticker with Minimum Weekly Return:")
print(min_weekly_return_row)

print("\nTicker with Maximum Weekly Return:")
print(max_weekly_return_row)

Ticker with Minimum Weekly Return:
                        Open   High        Low      Close  Adj Close  \
Date       ticker                                                      
2020-03-13 OKE     31.129999  31.34  26.870001  30.719999  22.432459   

                     Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                         
2020-03-13 OKE     14265000       0.07941     -0.192217               0   

                   Weekly_Return  Weekly_Ret_Class  
Date       ticker                                   
2020-03-13 OKE         -0.498121                 0  

Ticker with Maximum Weekly Return:
                       Open       High        Low      Close  Adj Close  \
Date       ticker                                                         
2020-03-26 PUK     24.72357  25.741999  24.548983  25.354025  24.007324   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker          

In [18]:
all_data.tail()

Open        High         Low       Close   Adj Close  \
Date       ticker                                                               
2024-07-15 ZTS     178.580002  180.850006  177.820007  177.979996  177.557648   
2024-07-16 ZTS     178.100006  183.220001  177.779999  182.910004  182.475967   
2024-07-17 ZTS     182.000000  184.839996  180.559998  182.050003  181.617996   
2024-07-18 ZTS     180.559998  184.240005  179.169998  179.229996  179.229996   
2024-07-19 ZTS     180.039993  180.509995  176.169998  178.580002  178.580002   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2024-07-15 ZTS     2216500     -0.005920      0.025880               1   
2024-07-16 ZTS     1939900      0.027700      0.029087               1   
2024-07-17 ZTS     2190800     -0.004702      0.016812               1   
2024-07-18 ZTS     2171100     -0.013148      0.009419               1   
2024-07-19 ZTS     2131000     -0.003627     -0.021351               0   

                   Weekly_Return  Weekly_Ret_Class  
Date       ticker                                   
2024-07-15 ZTS          0.014536                 1  
2024-07-16 ZTS          0.049759                 1  
2024-07-17 ZTS          0.049340                 1  
2024-07-18 ZTS          0.010782                 1  
2024-07-19 ZTS         -0.000197                 0

In [19]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1007902 entries, (Timestamp('2019-07-29 00:00:00'), 'A') to (Timestamp('2024-07-19 00:00:00'), 'ZTS')
Data columns (total 11 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   Open              1007902 non-null  float64
 1   High              1007902 non-null  float64
 2   Low               1007902 non-null  float64
 3   Close             1007902 non-null  float64
 4   Adj Close         1007902 non-null  float64
 5   Volume            1007902 non-null  int64  
 6   Daily_Return      1007902 non-null  float64
 7   3_Day_Return      1007902 non-null  float64
 8   3Day_Ret_Class    1007902 non-null  int64  
 9   Weekly_Return     1007902 non-null  float64
 10  Weekly_Ret_Class  1007902 non-null  int64  
dtypes: float64(8), int64(3)
memory usage: 88.5+ MB


In [20]:
import talib

# Calculate Relative Strength Index (RSI)
def calculate_rsi(df, period=14):
    df['RSI'] = talib.RSI(df['Adj Close'], timeperiod=period)
    return df

# Calculate Average Directional Index (ADX)
def calculate_adx(df, period=14):
    df['ADX'] = talib.ADX(df['High'], df['Low'], df['Close'], timeperiod=period)
    return df

# Apply the calculations to each group of tickers
all_data = all_data.groupby('ticker', group_keys=False).apply(calculate_rsi)
all_data = all_data.groupby('ticker', group_keys=False).apply(calculate_adx)

# Display the resulting DataFrame to verify the calculations
all_data.head()


Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
2019-07-30 A       69.250000  70.180000  68.099998  70.150002  67.736710   
2019-07-31 A       70.190002  70.660004  68.919998  69.410004  67.022156   
2019-08-01 A       69.629997  70.680000  68.389999  69.209999  66.829041   
2019-08-02 A       69.120003  69.129997  67.559998  68.080002  65.737915   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2019-07-29 A       2750000      0.012518      0.011354               1   
2019-07-30 A       2871800     -0.003126      0.004295               1   
2019-07-31 A       3846100     -0.010549     -0.001295               0   
2019-08-01 A       2494100     -0.002881     -0.016484               0   
2019-08-02 A       2803500     -0.016327     -0.029508               0   

                   Weekly_Return  Weekly_Ret_Class  RSI  ADX  
Date       ticker                                             
2019-07-29 A            0.018527                 1  NaN  NaN  
2019-07-30 A            0.003720                 1  NaN  NaN  
2019-07-31 A           -0.002443                 0  NaN  NaN  
2019-08-01 A           -0.009162                 0  NaN  NaN  
2019-08-02 A           -0.020432                 0  NaN  NaN

In [21]:
# Reset the index to include only 'ticker' for filling NaNs correctly
all_data.reset_index(inplace=True)
all_data.set_index('ticker', inplace=True)

# Handle NaNs: forward fill, then backward fill within each ticker group
all_data = all_data.groupby('ticker').ffill()
all_data = all_data.groupby('ticker').bfill()

In [22]:
all_data.head()

Date       Open       High        Low      Close  Adj Close  \
ticker                                                                     
A      2019-07-29  69.500000  70.500000  69.379997  70.370003  67.949135   
A      2019-07-30  69.250000  70.180000  68.099998  70.150002  67.736710   
A      2019-07-31  70.190002  70.660004  68.919998  69.410004  67.022156   
A      2019-08-01  69.629997  70.680000  68.389999  69.209999  66.829041   
A      2019-08-02  69.120003  69.129997  67.559998  68.080002  65.737915   

         Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  Weekly_Return  \
ticker                                                                       
A       2750000      0.012518      0.011354               1       0.018527   
A       2871800     -0.003126      0.004295               1       0.003720   
A       3846100     -0.010549     -0.001295               0      -0.002443   
A       2494100     -0.002881     -0.016484               0      -0.009162   
A       2803500     -0.016327     -0.029508               0      -0.020432   

        Weekly_Ret_Class        RSI       ADX  
ticker                                         
A                      1  51.815825  7.580604  
A                      1  51.815825  7.580604  
A                      0  51.815825  7.580604  
A                      0  51.815825  7.580604  
A                      0  51.815825  7.580604

In [23]:
# Convert the Date column to a DateTime index and set the multi-index
all_data.reset_index(inplace=True)
all_data['Date'] = pd.to_datetime(all_data['Date'])
all_data.set_index(['Date', 'ticker'], inplace=True)

# Display the resulting DataFrame to verify the multi-index structure
all_data.head()

Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
2019-07-30 A       69.250000  70.180000  68.099998  70.150002  67.736710   
2019-07-31 A       70.190002  70.660004  68.919998  69.410004  67.022156   
2019-08-01 A       69.629997  70.680000  68.389999  69.209999  66.829041   
2019-08-02 A       69.120003  69.129997  67.559998  68.080002  65.737915   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2019-07-29 A       2750000      0.012518      0.011354               1   
2019-07-30 A       2871800     -0.003126      0.004295               1   
2019-07-31 A       3846100     -0.010549     -0.001295               0   
2019-08-01 A       2494100     -0.002881     -0.016484               0   
2019-08-02 A       2803500     -0.016327     -0.029508               0   

                   Weekly_Return  Weekly_Ret_Class        RSI       ADX  
Date       ticker                                                        
2019-07-29 A            0.018527                 1  51.815825  7.580604  
2019-07-30 A            0.003720                 1  51.815825  7.580604  
2019-07-31 A           -0.002443                 0  51.815825  7.580604  
2019-08-01 A           -0.009162                 0  51.815825  7.580604  
2019-08-02 A           -0.020432                 0  51.815825  7.580604

In [24]:
import pandas as pd
import numpy as np

# Assuming `all_data` is your DataFrame

# Step 1: Generate lagged columns
for lag in range(1, 9):
    all_data[f'Weekly_Return_Lag_{lag}'] = all_data.groupby('ticker')['Weekly_Return'].shift(lag)

# Step 2: Handle NaNs in lagged columns
lag_columns = [f'Weekly_Return_Lag_{lag}' for lag in range(1, 6)]

# Forward fill NaNs
all_data[lag_columns] = all_data[lag_columns].ffill()

# Backward fill any remaining NaNs
all_data[lag_columns] = all_data[lag_columns].bfill()

all_data.head()

Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
2019-07-30 A       69.250000  70.180000  68.099998  70.150002  67.736710   
2019-07-31 A       70.190002  70.660004  68.919998  69.410004  67.022156   
2019-08-01 A       69.629997  70.680000  68.389999  69.209999  66.829041   
2019-08-02 A       69.120003  69.129997  67.559998  68.080002  65.737915   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2019-07-29 A       2750000      0.012518      0.011354               1   
2019-07-30 A       2871800     -0.003126      0.004295               1   
2019-07-31 A       3846100     -0.010549     -0.001295               0   
2019-08-01 A       2494100     -0.002881     -0.016484               0   
2019-08-02 A       2803500     -0.016327     -0.029508               0   

                   Weekly_Return  ...        RSI       ADX  \
Date       ticker                 ...                        
2019-07-29 A            0.018527  ...  51.815825  7.580604   
2019-07-30 A            0.003720  ...  51.815825  7.580604   
2019-07-31 A           -0.002443  ...  51.815825  7.580604   
2019-08-01 A           -0.009162  ...  51.815825  7.580604   
2019-08-02 A           -0.020432  ...  51.815825  7.580604   

                   Weekly_Return_Lag_1  Weekly_Return_Lag_2  \
Date       ticker                                             
2019-07-29 A                  0.018527             0.018527   
2019-07-30 A                  0.018527             0.018527   
2019-07-31 A                  0.003720             0.018527   
2019-08-01 A                 -0.002443             0.003720   
2019-08-02 A                 -0.009162            -0.002443   

                   Weekly_Return_Lag_3  Weekly_Return_Lag_4  \
Date       ticker                                             
2019-07-29 A                  0.018527             0.018527   
2019-07-30 A                  0.018527             0.018527   
2019-07-31 A                  0.018527             0.018527   
2019-08-01 A                  0.018527             0.018527   
2019-08-02 A                  0.003720             0.018527   

                   Weekly_Return_Lag_5  Weekly_Return_Lag_6  \
Date       ticker                                             
2019-07-29 A                  0.018527                  NaN   
2019-07-30 A                  0.018527                  NaN   
2019-07-31 A                  0.018527                  NaN   
2019-08-01 A                  0.018527                  NaN   
2019-08-02 A                  0.018527                  NaN   

                   Weekly_Return_Lag_7  Weekly_Return_Lag_8  
Date       ticker                                            
2019-07-29 A                       NaN                  NaN  
2019-07-30 A                       NaN                  NaN  
2019-07-31 A                       NaN                  NaN  
2019-08-01 A                       NaN                  NaN  
2019-08-02 A                       NaN                  NaN  

[5 rows x 21 columns]

In [25]:
all_data.tail()

Open        High         Low       Close   Adj Close  \
Date       ticker                                                               
2024-07-15 ZTS     178.580002  180.850006  177.820007  177.979996  177.557648   
2024-07-16 ZTS     178.100006  183.220001  177.779999  182.910004  182.475967   
2024-07-17 ZTS     182.000000  184.839996  180.559998  182.050003  181.617996   
2024-07-18 ZTS     180.559998  184.240005  179.169998  179.229996  179.229996   
2024-07-19 ZTS     180.039993  180.509995  176.169998  178.580002  178.580002   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2024-07-15 ZTS     2216500     -0.005920      0.025880               1   
2024-07-16 ZTS     1939900      0.027700      0.029087               1   
2024-07-17 ZTS     2190800     -0.004702      0.016812               1   
2024-07-18 ZTS     2171100     -0.013148      0.009419               1   
2024-07-19 ZTS     2131000     -0.003627     -0.021351               0   

                   Weekly_Return  ...        RSI        ADX  \
Date       ticker                 ...                         
2024-07-15 ZTS          0.014536  ...  58.997211  18.041663   
2024-07-16 ZTS          0.049759  ...  66.218288  19.009131   
2024-07-17 ZTS          0.049340  ...  64.097627  20.209661   
2024-07-18 ZTS          0.010782  ...  58.483610  20.800103   
2024-07-19 ZTS         -0.000197  ...  57.019694  20.324192   

                   Weekly_Return_Lag_1  Weekly_Return_Lag_2  \
Date       ticker                                             
2024-07-15 ZTS                0.023319             0.012187   
2024-07-16 ZTS                0.014536             0.023319   
2024-07-17 ZTS                0.049759             0.014536   
2024-07-18 ZTS                0.049340             0.049759   
2024-07-19 ZTS                0.010782             0.049340   

                   Weekly_Return_Lag_3  Weekly_Return_Lag_4  \
Date       ticker                                             
2024-07-15 ZTS               -0.007551             0.009619   
2024-07-16 ZTS                0.012187            -0.007551   
2024-07-17 ZTS                0.023319             0.012187   
2024-07-18 ZTS                0.014536             0.023319   
2024-07-19 ZTS                0.049759             0.014536   

                   Weekly_Return_Lag_5  Weekly_Return_Lag_6  \
Date       ticker                                             
2024-07-15 ZTS                0.011940            -0.008501   
2024-07-16 ZTS                0.009619             0.011940   
2024-07-17 ZTS               -0.007551             0.009619   
2024-07-18 ZTS                0.012187            -0.007551   
2024-07-19 ZTS                0.023319             0.012187   

                   Weekly_Return_Lag_7  Weekly_Return_Lag_8  
Date       ticker                                            
2024-07-15 ZTS                0.027502             0.040660  
2024-07-16 ZTS               -0.008501             0.027502  
2024-07-17 ZTS                0.011940            -0.008501  
2024-07-18 ZTS                0.009619             0.011940  
2024-07-19 ZTS               -0.007551             0.009619  

[5 rows x 21 columns]

In [26]:
import numpy as np

# Calculate Weekly Volatility
# Assuming a week has 5 trading days
all_data['Weekly_Volatility'] = all_data.groupby('ticker')['Daily_Return'].rolling(window=5).std().reset_index(level=0, drop=True)

# Calculate Weekly Relative Volume (RVOL)
all_data['Weekly_Volume'] = all_data.groupby('ticker')['Volume'].rolling(window=5).sum().reset_index(level=0, drop=True)
all_data['Average_Weekly_Volume'] = all_data.groupby('ticker')['Weekly_Volume'].rolling(window=4).mean().reset_index(level=0, drop=True)
all_data['Weekly_RVOL'] = all_data['Weekly_Volume'] / all_data['Average_Weekly_Volume']

# Calculate Weekly Rate of Change (ROC)
all_data['Weekly_ROC'] = all_data.groupby('ticker')['Close'].pct_change(periods=5)

# Drop intermediate columns to clean up the dataframe
all_data.drop(columns=['Weekly_Volume', 'Average_Weekly_Volume'], inplace=True)

# Display the first few rows to check the new columns
print(all_data.head())


                        Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
2019-07-30 A       69.250000  70.180000  68.099998  70.150002  67.736710   
2019-07-31 A       70.190002  70.660004  68.919998  69.410004  67.022156   
2019-08-01 A       69.629997  70.680000  68.389999  69.209999  66.829041   
2019-08-02 A       69.120003  69.129997  67.559998  68.080002  65.737915   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2019-07-29 A       2750000      0.012518      0.011354               1   
2019-07-30 A       2871800     -0.003126      0.004295               1   
2019-07-31 A       3846100     -0.010549     -0.001295               0   
2019-08-01 A       2494100     -0.002881     -0.016484               0   
2019-08-02 A       2803

In [27]:
# Handle NaNs using forward fill followed by backward fill
new_columns = ['Weekly_Volatility', 'Weekly_RVOL', 'Weekly_ROC']
all_data[new_columns] = all_data[new_columns].ffill().bfill()

In [28]:
all_data.head()

Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
2019-07-30 A       69.250000  70.180000  68.099998  70.150002  67.736710   
2019-07-31 A       70.190002  70.660004  68.919998  69.410004  67.022156   
2019-08-01 A       69.629997  70.680000  68.389999  69.209999  66.829041   
2019-08-02 A       69.120003  69.129997  67.559998  68.080002  65.737915   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2019-07-29 A       2750000      0.012518      0.011354               1   
2019-07-30 A       2871800     -0.003126      0.004295               1   
2019-07-31 A       3846100     -0.010549     -0.001295               0   
2019-08-01 A       2494100     -0.002881     -0.016484               0   
2019-08-02 A       2803500     -0.016327     -0.029508               0   

                   Weekly_Return  ...  Weekly_Return_Lag_2  \
Date       ticker                 ...                        
2019-07-29 A            0.018527  ...             0.018527   
2019-07-30 A            0.003720  ...             0.018527   
2019-07-31 A           -0.002443  ...             0.018527   
2019-08-01 A           -0.009162  ...             0.003720   
2019-08-02 A           -0.020432  ...            -0.002443   

                   Weekly_Return_Lag_3  Weekly_Return_Lag_4  \
Date       ticker                                             
2019-07-29 A                  0.018527             0.018527   
2019-07-30 A                  0.018527             0.018527   
2019-07-31 A                  0.018527             0.018527   
2019-08-01 A                  0.018527             0.018527   
2019-08-02 A                  0.003720             0.018527   

                   Weekly_Return_Lag_5  Weekly_Return_Lag_6  \
Date       ticker                                             
2019-07-29 A                  0.018527                  NaN   
2019-07-30 A                  0.018527                  NaN   
2019-07-31 A                  0.018527                  NaN   
2019-08-01 A                  0.018527                  NaN   
2019-08-02 A                  0.018527                  NaN   

                   Weekly_Return_Lag_7  Weekly_Return_Lag_8  \
Date       ticker                                             
2019-07-29 A                       NaN                  NaN   
2019-07-30 A                       NaN                  NaN   
2019-07-31 A                       NaN                  NaN   
2019-08-01 A                       NaN                  NaN   
2019-08-02 A                       NaN                  NaN   

                   Weekly_Volatility  Weekly_RVOL  Weekly_ROC  
Date       ticker                                              
2019-07-29 A                0.010836     0.907592    -0.06409  
2019-07-30 A                0.010836     0.907592    -0.06409  
2019-07-31 A                0.010836     0.907592    -0.06409  
2019-08-01 A                0.010836     0.907592    -0.06409  
2019-08-02 A                0.010836     0.907592    -0.06409  

[5 rows x 24 columns]

In [29]:
all_data.tail()

Open        High         Low       Close   Adj Close  \
Date       ticker                                                               
2024-07-15 ZTS     178.580002  180.850006  177.820007  177.979996  177.557648   
2024-07-16 ZTS     178.100006  183.220001  177.779999  182.910004  182.475967   
2024-07-17 ZTS     182.000000  184.839996  180.559998  182.050003  181.617996   
2024-07-18 ZTS     180.559998  184.240005  179.169998  179.229996  179.229996   
2024-07-19 ZTS     180.039993  180.509995  176.169998  178.580002  178.580002   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2024-07-15 ZTS     2216500     -0.005920      0.025880               1   
2024-07-16 ZTS     1939900      0.027700      0.029087               1   
2024-07-17 ZTS     2190800     -0.004702      0.016812               1   
2024-07-18 ZTS     2171100     -0.013148      0.009419               1   
2024-07-19 ZTS     2131000     -0.003627     -0.021351               0   

                   Weekly_Return  ...  Weekly_Return_Lag_2  \
Date       ticker                 ...                        
2024-07-15 ZTS          0.014536  ...             0.012187   
2024-07-16 ZTS          0.049759  ...             0.023319   
2024-07-17 ZTS          0.049340  ...             0.014536   
2024-07-18 ZTS          0.010782  ...             0.049759   
2024-07-19 ZTS         -0.000197  ...             0.049340   

                   Weekly_Return_Lag_3  Weekly_Return_Lag_4  \
Date       ticker                                             
2024-07-15 ZTS               -0.007551             0.009619   
2024-07-16 ZTS                0.012187            -0.007551   
2024-07-17 ZTS                0.023319             0.012187   
2024-07-18 ZTS                0.014536             0.023319   
2024-07-19 ZTS                0.049759             0.014536   

                   Weekly_Return_Lag_5  Weekly_Return_Lag_6  \
Date       ticker                                             
2024-07-15 ZTS                0.011940            -0.008501   
2024-07-16 ZTS                0.009619             0.011940   
2024-07-17 ZTS               -0.007551             0.009619   
2024-07-18 ZTS                0.012187            -0.007551   
2024-07-19 ZTS                0.023319             0.012187   

                   Weekly_Return_Lag_7  Weekly_Return_Lag_8  \
Date       ticker                                             
2024-07-15 ZTS                0.027502             0.040660   
2024-07-16 ZTS               -0.008501             0.027502   
2024-07-17 ZTS                0.011940            -0.008501   
2024-07-18 ZTS                0.009619             0.011940   
2024-07-19 ZTS               -0.007551             0.009619   

                   Weekly_Volatility  Weekly_RVOL  Weekly_ROC  
Date       ticker                                              
2024-07-15 ZTS              0.013317     1.054631    0.014536  
2024-07-16 ZTS              0.015721     1.036294    0.049759  
2024-07-17 ZTS              0.015812     1.038813    0.049340  
2024-07-18 ZTS              0.016012     1.028916    0.008383  
2024-07-19 ZTS              0.015894     1.048149   -0.002569  

[5 rows x 24 columns]

In [30]:
# Save the fixed DataFrame to a new CSV file, preserving the multi-index
all_data.to_csv('/home/annalisa-aaron/QuantProject/data/mega_historical_data.csv', index=True)

In [31]:
# Load the CSV file, preserving the Date and ticker multi-index
all_data = pd.read_csv('mega_historical_data.csv', index_col=['Date', 'ticker'], parse_dates=['Date'])

# Display the resulting DataFrame to verify the multi-index structure
all_data.head()

Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
2019-07-30 A       69.250000  70.180000  68.099998  70.150002  67.736710   
2019-07-31 A       70.190002  70.660004  68.919998  69.410004  67.022156   
2019-08-01 A       69.629997  70.680000  68.389999  69.209999  66.829041   
2019-08-02 A       69.120003  69.129997  67.559998  68.080002  65.737915   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2019-07-29 A       2750000      0.012518      0.011354               1   
2019-07-30 A       2871800     -0.003126      0.004295               1   
2019-07-31 A       3846100     -0.010549     -0.001295               0   
2019-08-01 A       2494100     -0.002881     -0.016484               0   
2019-08-02 A       2803500     -0.016327     -0.029508               0   

                   Weekly_Return  ...  Weekly_Return_Lag_2  \
Date       ticker                 ...                        
2019-07-29 A            0.018527  ...             0.018527   
2019-07-30 A            0.003720  ...             0.018527   
2019-07-31 A           -0.002443  ...             0.018527   
2019-08-01 A           -0.009162  ...             0.003720   
2019-08-02 A           -0.020432  ...            -0.002443   

                   Weekly_Return_Lag_3  Weekly_Return_Lag_4  \
Date       ticker                                             
2019-07-29 A                  0.018527             0.018527   
2019-07-30 A                  0.018527             0.018527   
2019-07-31 A                  0.018527             0.018527   
2019-08-01 A                  0.018527             0.018527   
2019-08-02 A                  0.003720             0.018527   

                   Weekly_Return_Lag_5  Weekly_Return_Lag_6  \
Date       ticker                                             
2019-07-29 A                  0.018527                  NaN   
2019-07-30 A                  0.018527                  NaN   
2019-07-31 A                  0.018527                  NaN   
2019-08-01 A                  0.018527                  NaN   
2019-08-02 A                  0.018527                  NaN   

                   Weekly_Return_Lag_7  Weekly_Return_Lag_8  \
Date       ticker                                             
2019-07-29 A                       NaN                  NaN   
2019-07-30 A                       NaN                  NaN   
2019-07-31 A                       NaN                  NaN   
2019-08-01 A                       NaN                  NaN   
2019-08-02 A                       NaN                  NaN   

                   Weekly_Volatility  Weekly_RVOL  Weekly_ROC  
Date       ticker                                              
2019-07-29 A                0.010836     0.907592    -0.06409  
2019-07-30 A                0.010836     0.907592    -0.06409  
2019-07-31 A                0.010836     0.907592    -0.06409  
2019-08-01 A                0.010836     0.907592    -0.06409  
2019-08-02 A                0.010836     0.907592    -0.06409  

[5 rows x 24 columns]

In [32]:
all_data.tail()

Open        High         Low       Close   Adj Close  \
Date       ticker                                                               
2024-07-15 ZTS     178.580002  180.850006  177.820007  177.979996  177.557648   
2024-07-16 ZTS     178.100006  183.220001  177.779999  182.910004  182.475967   
2024-07-17 ZTS     182.000000  184.839996  180.559998  182.050003  181.617996   
2024-07-18 ZTS     180.559998  184.240005  179.169998  179.229996  179.229996   
2024-07-19 ZTS     180.039993  180.509995  176.169998  178.580002  178.580002   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2024-07-15 ZTS     2216500     -0.005920      0.025880               1   
2024-07-16 ZTS     1939900      0.027700      0.029087               1   
2024-07-17 ZTS     2190800     -0.004702      0.016812               1   
2024-07-18 ZTS     2171100     -0.013148      0.009419               1   
2024-07-19 ZTS     2131000     -0.003627     -0.021351               0   

                   Weekly_Return  ...  Weekly_Return_Lag_2  \
Date       ticker                 ...                        
2024-07-15 ZTS          0.014536  ...             0.012187   
2024-07-16 ZTS          0.049759  ...             0.023319   
2024-07-17 ZTS          0.049340  ...             0.014536   
2024-07-18 ZTS          0.010782  ...             0.049759   
2024-07-19 ZTS         -0.000197  ...             0.049340   

                   Weekly_Return_Lag_3  Weekly_Return_Lag_4  \
Date       ticker                                             
2024-07-15 ZTS               -0.007551             0.009619   
2024-07-16 ZTS                0.012187            -0.007551   
2024-07-17 ZTS                0.023319             0.012187   
2024-07-18 ZTS                0.014536             0.023319   
2024-07-19 ZTS                0.049759             0.014536   

                   Weekly_Return_Lag_5  Weekly_Return_Lag_6  \
Date       ticker                                             
2024-07-15 ZTS                0.011940            -0.008501   
2024-07-16 ZTS                0.009619             0.011940   
2024-07-17 ZTS               -0.007551             0.009619   
2024-07-18 ZTS                0.012187            -0.007551   
2024-07-19 ZTS                0.023319             0.012187   

                   Weekly_Return_Lag_7  Weekly_Return_Lag_8  \
Date       ticker                                             
2024-07-15 ZTS                0.027502             0.040660   
2024-07-16 ZTS               -0.008501             0.027502   
2024-07-17 ZTS                0.011940            -0.008501   
2024-07-18 ZTS                0.009619             0.011940   
2024-07-19 ZTS               -0.007551             0.009619   

                   Weekly_Volatility  Weekly_RVOL  Weekly_ROC  
Date       ticker                                              
2024-07-15 ZTS              0.013317     1.054631    0.014536  
2024-07-16 ZTS              0.015721     1.036294    0.049759  
2024-07-17 ZTS              0.015812     1.038813    0.049340  
2024-07-18 ZTS              0.016012     1.028916    0.008383  
2024-07-19 ZTS              0.015894     1.048149   -0.002569  

[5 rows x 24 columns]

In [33]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1007902 entries, (Timestamp('2019-07-29 00:00:00'), 'A') to (Timestamp('2024-07-19 00:00:00'), 'ZTS')
Data columns (total 24 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   Open                 1007902 non-null  float64
 1   High                 1007902 non-null  float64
 2   Low                  1007902 non-null  float64
 3   Close                1007902 non-null  float64
 4   Adj Close            1007902 non-null  float64
 5   Volume               1007902 non-null  int64  
 6   Daily_Return         1007902 non-null  float64
 7   3_Day_Return         1007902 non-null  float64
 8   3Day_Ret_Class       1007902 non-null  int64  
 9   Weekly_Return        1007902 non-null  float64
 10  Weekly_Ret_Class     1007902 non-null  int64  
 11  RSI                  1007902 non-null  float64
 12  ADX                  1007902 non-null  float64
 13  Weekly_Return_Lag_1  1007902 n

In [34]:
import pandas as pd
import numpy as np

def ewma_half_life(data, span):
    alpha = 1 - np.exp(np.log(0.5) / span)
    return data.ewm(alpha=alpha, adjust=False).mean()

# Step 1: Calculate the EWMA for different time scales
short_scales = [8, 16, 32]
long_scales = [24, 48, 96]

for s in short_scales:
    all_data[f'EWMA_{s}'] = all_data.groupby('ticker')['Close'].transform(lambda x: ewma_half_life(x, s))

for l in long_scales:
    all_data[f'EWMA_{l}'] = all_data.groupby('ticker')['Close'].transform(lambda x: ewma_half_life(x, l))

# Step 2: Calculate the 63-day rolling standard deviation
all_data['Rolling_Std_63'] = all_data.groupby('ticker')['Close'].transform(lambda x: x.rolling(window=63).std())

# Step 3: Compute the volatility-scaled MACD
for s in short_scales:
    for l in long_scales:
        if l > s:
            all_data[f'MACD_{s}_{l}'] = (all_data[f'EWMA_{s}'] - all_data[f'EWMA_{l}']) / all_data['Rolling_Std_63']

# Step 4: Combine the volatility-scaled MACDs into a composite signal
macd_columns = [f'MACD_{s}_{l}' for s in short_scales for l in long_scales if l > s]
all_data['Composite_MACD'] = all_data[macd_columns].sum(axis=1)

# Handle NaNs using forward fill followed by backward fill
all_data.ffill(inplace=True)
all_data.bfill(inplace=True)

# Display the first few rows to check the new columns
print(all_data.head())


                        Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
2019-07-30 A       69.250000  70.180000  68.099998  70.150002  67.736710   
2019-07-31 A       70.190002  70.660004  68.919998  69.410004  67.022156   
2019-08-01 A       69.629997  70.680000  68.389999  69.209999  66.829041   
2019-08-02 A       69.120003  69.129997  67.559998  68.080002  65.737915   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2019-07-29 A       2750000      0.012518      0.011354               1   
2019-07-30 A       2871800     -0.003126      0.004295               1   
2019-07-31 A       3846100     -0.010549     -0.001295               0   
2019-08-01 A       2494100     -0.002881     -0.016484               0   
2019-08-02 A       2803

In [35]:
all_data.tail()

Open        High         Low       Close   Adj Close  \
Date       ticker                                                               
2024-07-15 ZTS     178.580002  180.850006  177.820007  177.979996  177.557648   
2024-07-16 ZTS     178.100006  183.220001  177.779999  182.910004  182.475967   
2024-07-17 ZTS     182.000000  184.839996  180.559998  182.050003  181.617996   
2024-07-18 ZTS     180.559998  184.240005  179.169998  179.229996  179.229996   
2024-07-19 ZTS     180.039993  180.509995  176.169998  178.580002  178.580002   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2024-07-15 ZTS     2216500     -0.005920      0.025880               1   
2024-07-16 ZTS     1939900      0.027700      0.029087               1   
2024-07-17 ZTS     2190800     -0.004702      0.016812               1   
2024-07-18 ZTS     2171100     -0.013148      0.009419               1   
2024-07-19 ZTS     2131000     -0.003627     -0.021351               0   

                   Weekly_Return  ...  Rolling_Std_63  MACD_8_24  MACD_8_48  \
Date       ticker                 ...                                         
2024-07-15 ZTS          0.014536  ...        8.725083   0.197248   0.025496   
2024-07-16 ZTS          0.049759  ...        8.586176   0.249982   0.095105   
2024-07-17 ZTS          0.049340  ...        8.501353   0.290943   0.152162   
2024-07-18 ZTS          0.010782  ...        8.294597   0.314075   0.184404   
2024-07-19 ZTS         -0.000197  ...        8.088811   0.331431   0.209721   

                   MACD_8_96  MACD_16_24  MACD_16_48  MACD_16_96  MACD_32_48  \
Date       ticker                                                              
2024-07-15 ZTS     -0.070910    0.016053   -0.155698   -0.252104   -0.117434   
2024-07-16 ZTS      0.005267    0.032526   -0.122351   -0.212190   -0.109408   
2024-07-17 ZTS      0.068746    0.046636   -0.092144   -0.175560   -0.101513   
2024-07-18 ZTS      0.103830    0.056151   -0.073521   -0.154095   -0.097539   
2024-07-19 ZTS      0.131470    0.064358   -0.057351   -0.135602   -0.094126   

                   MACD_32_96  Composite_MACD  
Date       ticker                              
2024-07-15 ZTS      -0.213839       -0.571188  
2024-07-16 ZTS      -0.199246       -0.260314  
2024-07-17 ZTS      -0.184929        0.004340  
2024-07-18 ZTS      -0.178113        0.155193  
2024-07-19 ZTS      -0.172377        0.277524  

[5 rows x 40 columns]

In [36]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1007902 entries, (Timestamp('2019-07-29 00:00:00'), 'A') to (Timestamp('2024-07-19 00:00:00'), 'ZTS')
Data columns (total 40 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   Open                 1007902 non-null  float64
 1   High                 1007902 non-null  float64
 2   Low                  1007902 non-null  float64
 3   Close                1007902 non-null  float64
 4   Adj Close            1007902 non-null  float64
 5   Volume               1007902 non-null  int64  
 6   Daily_Return         1007902 non-null  float64
 7   3_Day_Return         1007902 non-null  float64
 8   3Day_Ret_Class       1007902 non-null  int64  
 9   Weekly_Return        1007902 non-null  float64
 10  Weekly_Ret_Class     1007902 non-null  int64  
 11  RSI                  1007902 non-null  float64
 12  ADX                  1007902 non-null  float64
 13  Weekly_Return_Lag_1  1007902 n

In [37]:
# Save the fixed DataFrame to a new CSV file, preserving the multi-index
all_data.to_csv('mega_historical_data2.csv', index=True)

In [38]:
# Reorder the multi-index to prioritize dates
all_data = all_data.reorder_levels(['Date', 'ticker']).sort_index(level='Date')

# Display the first few rows to check the new index order
print(all_data.head())


                        Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
           AAPL    52.115002  52.660000  52.110001  52.419998  50.590385   
           ABBV    67.870003  68.410004  66.879997  67.180000  53.860332   
           ABEV     5.230000   5.430000   5.200000   5.400000   4.422829   
           ABT     88.059998  88.620003  87.379997  88.550003  81.213249   

                     Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                         
2019-07-29 A        2750000      0.012518      0.011354               1   
           AAPL    86693600      0.009338      0.004840               1   
           ABBV    10117300     -0.008560     -0.004151               0   
           ABEV    35419600      0.030534      0.141649               1   
           ABT   

In [39]:
# Apply ranking by date to the Weekly_Return column and Weekly_Return_Lag_1 column

# Rank Weekly_Return within each date
all_data['Weekly_Return_Rank'] = all_data.groupby('Date')['Weekly_Return'].rank(ascending=False)

# Rank Weekly_Return_Lag_1 within each date
all_data['Weekly_Return_Lag_1_Rank'] = all_data.groupby('Date')['Weekly_Return_Lag_1'].rank(ascending=False)

# Display the first few rows to check the new ranking columns
print(all_data.head())


                        Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
           AAPL    52.115002  52.660000  52.110001  52.419998  50.590385   
           ABBV    67.870003  68.410004  66.879997  67.180000  53.860332   
           ABEV     5.230000   5.430000   5.200000   5.400000   4.422829   
           ABT     88.059998  88.620003  87.379997  88.550003  81.213249   

                     Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                         
2019-07-29 A        2750000      0.012518      0.011354               1   
           AAPL    86693600      0.009338      0.004840               1   
           ABBV    10117300     -0.008560     -0.004151               0   
           ABEV    35419600      0.030534      0.141649               1   
           ABT   

In [40]:
all_data.tail()

Open        High         Low       Close   Adj Close  \
Date       ticker                                                               
2024-07-19 ZG       49.720001   49.970001   48.810001   49.099998   49.099998   
           ZM       58.830002   59.090000   58.270000   58.869999   58.869999   
           ZS      188.960007  191.970001  186.500000  188.539993  188.539993   
           ZTO      19.940001   20.000000   19.702000   19.750000   19.750000   
           ZTS     180.039993  180.509995  176.169998  178.580002  178.580002   

                    Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                        
2024-07-19 ZG       402300     -0.011277     -0.066185               0   
           ZM      1840800     -0.004734     -0.004902               0   
           ZS      1623700     -0.009613     -0.071780               0   
           ZTO     2099400     -0.012994     -0.024209               0   
           ZTS     2131000     -0.003627     -0.021351               0   

                   Weekly_Return  ...  MACD_8_48  MACD_8_96  MACD_16_24  \
Date       ticker                 ...                                     
2024-07-19 ZG           0.004295  ...   0.319783   0.228422    0.050794   
           ZM           0.022404  ...  -1.693778  -3.977864   -0.394419   
           ZS          -0.066171  ...   0.313724   0.879000    0.011024   
           ZTO         -0.036115  ...  -0.446416  -1.022235   -0.045687   
           ZTS         -0.000197  ...   0.209721   0.131470    0.064358   

                   MACD_16_48  MACD_16_96  MACD_32_48  MACD_32_96  \
Date       ticker                                                   
2024-07-19 ZG       -0.082209   -0.173570   -0.089271   -0.180632   
           ZM       -1.263316   -3.547402   -0.537414   -2.821500   
           ZS       -0.031661    0.533615    0.015164    0.580440   
           ZTO      -0.131919   -0.707739   -0.082547   -0.658366   
           ZTS      -0.057351   -0.135602   -0.094126   -0.172377   

                   Composite_MACD  Weekly_Return_Rank  \
Date       ticker                                       
2024-07-19 ZG            0.526102               335.0   
           ZM          -15.060574               163.0   
           ZS            2.657717               756.0   
           ZTO          -3.455092               674.0   
           ZTS           0.277524               392.0   

                   Weekly_Return_Lag_1_Rank  
Date       ticker                            
2024-07-19 ZG                         244.0  
           ZM                         110.0  
           ZS                         688.0  
           ZTO                        602.0  
           ZTS                        420.0  

[5 rows x 42 columns]

In [41]:
# Select the specific date you want to inspect
specific_date = '2024-07-19'  # Replace with the date you want to check

# Extract Weekly_Return and Weekly_Return_Rank for the specified date and sort by rank
data_for_date = all_data.loc[specific_date, ['Weekly_Return', 'Weekly_Return_Rank', 'Weekly_Return_Lag_1', 'Weekly_Return_Lag_1_Rank']]
data_for_date_sorted = data_for_date.sort_values(by='Weekly_Return_Rank')

# Extract top 10 and bottom 10
top_10 = data_for_date_sorted.head(10)
bottom_10 = data_for_date_sorted.tail(10)

# Display the top 10 and bottom 10
print("Top 10 Tickers by Weekly Return Rank on", specific_date)
print(top_10)
print("\nBottom 10 Tickers by Weekly Return Rank on", specific_date)
print(bottom_10)


Top 10 Tickers by Weekly Return Rank on 2024-07-19
        Weekly_Return  Weekly_Return_Rank  Weekly_Return_Lag_1  \
ticker                                                           
MSTR         0.273304                 1.0             0.144712   
COIN         0.182414                 2.0             0.086707   
WBD          0.171622                 3.0             0.154472   
DHI          0.130740                 4.0             0.158218   
UNH          0.105175                 5.0             0.118923   
HBAN         0.097489                 6.0             0.062407   
CBOE         0.089263                 7.0             0.083466   
OWL          0.088701                 8.0             0.089347   
DFS          0.079906                 9.0             0.086781   
EPAM         0.079587                10.0             0.149601   

        Weekly_Return_Lag_1_Rank  
ticker                            
MSTR                         4.0  
COIN                        20.0  
WBD             

In [42]:
# Select the specific date you want to inspect
specific_date = '2024-07-19'  # Replace with the date you want to check

# Extract Weekly_Return and Weekly_Return_Rank for the specified date and sort by rank
data_for_date = all_data.loc[specific_date, ['Weekly_Return', 'Weekly_Return_Rank', 'Weekly_Return_Lag_1', 'Weekly_Return_Lag_1_Rank']]
data_for_date_sorted = data_for_date.sort_values(by='Weekly_Return_Rank')

# Extract top 10 and bottom 10
top_10 = data_for_date_sorted.head(10)
bottom_10 = data_for_date_sorted.tail(10)

# Display the top 10 and bottom 10
print("Top 10 Tickers by Weekly Return Rank on", specific_date)
print(top_10)
print("\nBottom 10 Tickers by Weekly Return Rank on", specific_date)
print(bottom_10)


Top 10 Tickers by Weekly Return Rank on 2024-07-19
        Weekly_Return  Weekly_Return_Rank  Weekly_Return_Lag_1  \
ticker                                                           
MSTR         0.273304                 1.0             0.144712   
COIN         0.182414                 2.0             0.086707   
WBD          0.171622                 3.0             0.154472   
DHI          0.130740                 4.0             0.158218   
UNH          0.105175                 5.0             0.118923   
HBAN         0.097489                 6.0             0.062407   
CBOE         0.089263                 7.0             0.083466   
OWL          0.088701                 8.0             0.089347   
DFS          0.079906                 9.0             0.086781   
EPAM         0.079587                10.0             0.149601   

        Weekly_Return_Lag_1_Rank  
ticker                            
MSTR                         4.0  
COIN                        20.0  
WBD             

In [43]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1007902 entries, (Timestamp('2019-07-29 00:00:00'), 'A') to (Timestamp('2024-07-19 00:00:00'), 'ZTS')
Data columns (total 42 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   Open                      1007902 non-null  float64
 1   High                      1007902 non-null  float64
 2   Low                       1007902 non-null  float64
 3   Close                     1007902 non-null  float64
 4   Adj Close                 1007902 non-null  float64
 5   Volume                    1007902 non-null  int64  
 6   Daily_Return              1007902 non-null  float64
 7   3_Day_Return              1007902 non-null  float64
 8   3Day_Ret_Class            1007902 non-null  int64  
 9   Weekly_Return             1007902 non-null  float64
 10  Weekly_Ret_Class          1007902 non-null  int64  
 11  RSI                       1007902 non-null  float64
 12  ADX         

In [44]:
# Save the fixed DataFrame to a new CSV file, preserving the multi-index
all_data.to_csv('mega_historical_data3.csv', index=True)

In [45]:
# Load the CSV file, preserving the Date and ticker multi-index
all_data = pd.read_csv('mega_historical_data3.csv', index_col=['Date', 'ticker'], parse_dates=['Date'])

In [46]:
all_data.head()

Open       High        Low      Close  Adj Close  \
Date       ticker                                                          
2019-07-29 A       69.500000  70.500000  69.379997  70.370003  67.949135   
           AAPL    52.115002  52.660000  52.110001  52.419998  50.590385   
           ABBV    67.870003  68.410004  66.879997  67.180000  53.860332   
           ABEV     5.230000   5.430000   5.200000   5.400000   4.422829   
           ABT     88.059998  88.620003  87.379997  88.550003  81.213249   

                     Volume  Daily_Return  3_Day_Return  3Day_Ret_Class  \
Date       ticker                                                         
2019-07-29 A        2750000      0.012518      0.011354               1   
           AAPL    86693600      0.009338      0.004840               1   
           ABBV    10117300     -0.008560     -0.004151               0   
           ABEV    35419600      0.030534      0.141649               1   
           ABT      3623900      0.009117     -0.000452               0   

                   Weekly_Return  ...  MACD_8_48  MACD_8_96  MACD_16_24  \
Date       ticker                 ...                                     
2019-07-29 A            0.018527  ...   0.763743   1.013300    0.205818   
           AAPL         0.011871  ...  -0.391316  -0.175650   -0.149005   
           ABBV        -0.012059  ...   1.321257   1.754209    0.334309   
           ABEV         0.132076  ...   0.805643   1.835905    0.101041   
           ABT          0.005450  ...   0.092175   1.126424   -0.084860   

                   MACD_16_48  MACD_16_96  MACD_32_48  MACD_32_96  \
Date       ticker                                                   
2019-07-29 A         0.555444    0.805001    0.196798    0.446355   
           AAPL     -0.120660    0.095006    0.065513    0.281179   
           ABBV      0.811726    1.244677    0.262141    0.695093   
           ABEV      0.561654    1.591916    0.339157    1.369419   
           ABT       0.213969    1.248217    0.270326    1.304574   

                   Composite_MACD  Weekly_Return_Rank  \
Date       ticker                                       
2019-07-29 A                  0.0               276.0   
           AAPL               0.0               359.0   
           ABBV               0.0               605.0   
           ABEV               0.0                10.0   
           ABT                0.0               430.0   

                   Weekly_Return_Lag_1_Rank  
Date       ticker                            
2019-07-29 A                          299.0  
           AAPL                       392.0  
           ABBV                       568.0  
           ABEV                       322.0  
           ABT                        417.0  

[5 rows x 42 columns]

In [47]:
# List of columns to drop
columns_to_drop = ['Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume', 
                   'EWMA_8', 'EWMA_16', 'EWMA_32', 'EWMA_24', 'EWMA_48', 'EWMA_96']

# Drop the specified columns
all_data.drop(columns=columns_to_drop, inplace=True)

# Display the first few rows to check the remaining columns
all_data.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1007902 entries, (Timestamp('2019-07-29 00:00:00'), 'A') to (Timestamp('2024-07-19 00:00:00'), 'ZTS')
Data columns (total 30 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   Daily_Return              1007902 non-null  float64
 1   3_Day_Return              1007902 non-null  float64
 2   3Day_Ret_Class            1007902 non-null  int64  
 3   Weekly_Return             1007902 non-null  float64
 4   Weekly_Ret_Class          1007902 non-null  int64  
 5   RSI                       1007902 non-null  float64
 6   ADX                       1007902 non-null  float64
 7   Weekly_Return_Lag_1       1007902 non-null  float64
 8   Weekly_Return_Lag_2       1007902 non-null  float64
 9   Weekly_Return_Lag_3       1007902 non-null  float64
 10  Weekly_Return_Lag_4       1007902 non-null  float64
 11  Weekly_Return_Lag_5       1007902 non-null  float64
 12  Weekly_Retur

In [48]:
# Save the fixed DataFrame to a new CSV file, preserving the multi-index
all_data.to_csv('mega_historical_data4.csv', index=True)

Code below to restore notebook. Cell code above experimental.

This model is a LambdaMART model that has a 1.16% weekly return. Should try running it with the inclusion of trend scanning as a feature, if possible.

In [49]:
import xgboost as xgb
import pandas as pd

# Step 1: Prepare the Data
# Assuming 'all_data' is your multi-index DataFrame with Date and ticker

# Flatten the multi-index for compatibility with XGBoost
all_data = all_data.reset_index()

# Define features and target
features = ['3Day_Ret_Class', 'Weekly_Ret_Class', 'RSI', 'ADX', 'Weekly_Return_Lag_1', 'Weekly_Return_Lag_2',
           'Weekly_Return_Lag_3', 'Weekly_Return_Lag_4', 'Weekly_Return_Lag_5', 'Weekly_Volatility', 
            'Weekly_RVOL', 'Weekly_ROC', 'MACD_8_24', 'MACD_16_24', 'MACD_16_96', 'Weekly_Return_Lag_1_Rank'
           ]

# Rank the weekly returns and rescale the ranks to fit within the range 0 to 31
all_data['Weekly_Return_Rank'] = all_data.groupby('Date')['Weekly_Return'].rank(method='first', ascending=False)
all_data['Weekly_Return_Rank'] = ((all_data['Weekly_Return_Rank'] - 1) * (31 / (all_data['Weekly_Return_Rank'].max() - 1))).astype(int)

# Use the ranks as the target
target = 'Weekly_Return_Rank'

# Create query ID column based on Date
all_data['qid'] = all_data['Date'].factorize()[0]

# Split the data into training and testing sets
train_data = all_data[all_data['Date'] < '2023-06-01']
test_data = all_data[all_data['Date'] >= '2023-06-01']

# Create DMatrix for XGBoost with query ID
dtrain = xgb.DMatrix(train_data[features], label=train_data[target])
dtrain.set_group(train_data.groupby('qid').size().to_list())

dtest = xgb.DMatrix(test_data[features], label=test_data[target])
dtest.set_group(test_data.groupby('qid').size().to_list())

# Step 2: Train the LambdaMART Model
params = {
    'objective': 'rank:pairwise',  # Pairwise ranking objective
    'eval_metric': 'ndcg',         # NDCG metric for evaluation
    'eta': 0.1,
    'gamma': 1.0,
    'min_child_weight': 0.1,
    'max_depth': 6,
    'nthread': 4,
    'tree_method': 'hist',
    'device': 'cuda'  # Use GPU acceleration if available
}

num_rounds = 100  # Number of boosting rounds
model = xgb.train(params, dtrain, num_rounds, evals=[(dtest, 'eval')], early_stopping_rounds=10)

# Step 3: Evaluate and Predict
predicted_scores = model.predict(dtest)

# Convert predicted scores to ranks (ascending order of scores)
test_data['Predicted_Score'] = predicted_scores
test_data['Predicted_Rank'] = test_data.groupby('Date')['Predicted_Score'].rank(method='first', ascending=True)

# Restore multi-index
test_data = test_data.set_index(['Date', 'ticker'])

# Select a specific date for inspection
specific_date = '2024-07-12'
data_for_date = test_data.xs(specific_date, level='Date')

# Display the top 10 and bottom 10
top_10 = data_for_date.nsmallest(10, 'Predicted_Rank')
bottom_10 = data_for_date.nlargest(10, 'Predicted_Rank')

print("Top 10 Tickers by Predicted Rank on", specific_date)
print(top_10[['Weekly_Return', 'Weekly_Return_Rank', 'Predicted_Score', 'Predicted_Rank']])

print("\nBottom 10 Tickers by Predicted Rank on", specific_date)
print(bottom_10[['Weekly_Return', 'Weekly_Return_Rank', 'Predicted_Score', 'Predicted_Rank']])


[0]	eval-ndcg:0.75151
[1]	eval-ndcg:0.90684
[2]	eval-ndcg:0.93133
[3]	eval-ndcg:0.93855
[4]	eval-ndcg:0.94114
[5]	eval-ndcg:0.94494
[6]	eval-ndcg:0.94841
[7]	eval-ndcg:0.95035
[8]	eval-ndcg:0.95232
[9]	eval-ndcg:0.95465
[10]	eval-ndcg:0.95656
[11]	eval-ndcg:0.95821
[12]	eval-ndcg:0.95893
[13]	eval-ndcg:0.95978
[14]	eval-ndcg:0.96095
[15]	eval-ndcg:0.96140
[16]	eval-ndcg:0.96138
[17]	eval-ndcg:0.96175
[18]	eval-ndcg:0.96221
[19]	eval-ndcg:0.96263
[20]	eval-ndcg:0.96274
[21]	eval-ndcg:0.96318
[22]	eval-ndcg:0.96368
[23]	eval-ndcg:0.96388
[24]	eval-ndcg:0.96410
[25]	eval-ndcg:0.96430
[26]	eval-ndcg:0.96463
[27]	eval-ndcg:0.96485
[28]	eval-ndcg:0.96536
[29]	eval-ndcg:0.96553
[30]	eval-ndcg:0.96576
[31]	eval-ndcg:0.96590
[32]	eval-ndcg:0.96669
[33]	eval-ndcg:0.96712
[34]	eval-ndcg:0.96735
[35]	eval-ndcg:0.96768
[36]	eval-ndcg:0.96789
[37]	eval-ndcg:0.96825
[38]	eval-ndcg:0.96848
[39]	eval-ndcg:0.96869
[40]	eval-ndcg:0.96877
[41]	eval-ndcg:0.96924
[42]	eval-ndcg:0.96946
[43]	eval-ndcg:0.9693

/tmp/ipykernel_240024/3886942321.py:57: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_data['Predicted_Score'] = predicted_scores
/tmp/ipykernel_240024/3886942321.py:58: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_data['Predicted_Rank'] = test_data.groupby('Date')['Predicted_Score'].rank(method='first', ascending=True)


In [50]:
# Select a specific date for inspection
specific_date = '2024-07-19'
data_for_date = test_data.xs(specific_date, level='Date')

# Display the top 10 and bottom 10
top_10 = data_for_date.nsmallest(10, 'Predicted_Rank')
bottom_10 = data_for_date.nlargest(10, 'Predicted_Rank')

print("Top 10 Tickers by Predicted Rank on", specific_date)
print(top_10[['Weekly_Return', 'Weekly_Return_Rank', 'Predicted_Score', 'Predicted_Rank']])

print("\nBottom 10 Tickers by Predicted Rank on", specific_date)
print(bottom_10[['Weekly_Return', 'Weekly_Return_Rank', 'Predicted_Score', 'Predicted_Rank']])


Top 10 Tickers by Predicted Rank on 2024-07-19
        Weekly_Return  Weekly_Return_Rank  Predicted_Score  Predicted_Rank
ticker                                                                    
ALL          0.065306                   0        -5.290235             1.0
APA          0.068068                   0        -5.290235             2.0
BX           0.071683                   0        -5.290235             3.0
CBOE         0.089263                   0        -5.290235             4.0
CFG          0.073298                   0        -5.290235             5.0
CG           0.062688                   1        -5.290235             6.0
COF          0.064013                   1        -5.290235             7.0
COIN         0.182414                   0        -5.290235             8.0
DFS          0.079906                   0        -5.290235             9.0
DHI          0.130740                   0        -5.290235            10.0

Bottom 10 Tickers by Predicted Rank on 2024-07-19
  

CODE TO DETERMINE EQUAL SHARES

In [60]:
import yfinance as yf

# List of top and bottom tickers on 2024-07-20
top_tickers = ['ALL', 'APA', 'BX', 'CBOE', 'CFG', 'CG', 'COF', 'COIN', 'DFS', 'DHI']
bottom_tickers = ['DPZ', 'ASML', 'CRWD', 'SCHW', 'LPLA', 'AMD', 'CELH', 'ONTO', 'VST', 'LRCX']

# Fetch data from yfinance for the adjusted closing prices on 2024-07-20
tickers = top_tickers + bottom_tickers
data = yf.download(tickers, start='2024-07-14', end='2024-07-20')

# Initialize the investment parameters
total_portfolio_value = 50000  # Total available in account
individual_investment = 2000  # Approximate investment per ticker

# Calculate the number of shares to buy for each ticker
shares_to_buy = {}
total_invested = 0

for ticker in tickers:
    adj_close_price = data['Adj Close'][ticker].iloc[0]
    num_shares = int(individual_investment // adj_close_price)
    # Ensure we invest as close to 2000 as possible without exceeding total account value
    if num_shares * adj_close_price < individual_investment:
        num_shares += 1
    investment = num_shares * adj_close_price
    total_invested += investment
    shares_to_buy[ticker] = num_shares

    # Adjust the investment if we're close to the total portfolio value
    if total_invested > total_portfolio_value:
        break

# Display the number of shares to buy for each ticker
print("Number of Shares to Buy for Top 10 Predicted Tickers on 2024-06-28:")
for ticker in top_tickers:
    if ticker in shares_to_buy:
        print(f"{ticker}: {shares_to_buy[ticker]} shares")

print("\nNumber of Shares to Sell for Bottom 10 Predicted Tickers on 2024-06-28:")
for ticker in bottom_tickers:
    if ticker in shares_to_buy:
        print(f"{ticker}: {shares_to_buy[ticker]} shares")

print(f"\nTotal Invested: ${total_invested:.2f}")


[*********************100%%**********************]  20 of 20 completed
Number of Shares to Buy for Top 10 Predicted Tickers on 2024-06-28:
ALL: 12 shares
APA: 64 shares
BX: 16 shares
CBOE: 12 shares
CFG: 52 shares
CG: 46 shares
COF: 14 shares
COIN: 9 shares
DFS: 15 shares
DHI: 14 shares

Number of Shares to Sell for Bottom 10 Predicted Tickers on 2024-06-28:
DPZ: 5 shares
ASML: 2 shares
CRWD: 6 shares
SCHW: 27 shares
LPLA: 8 shares
AMD: 12 shares
CELH: 39 shares
ONTO: 9 shares
VST: 24 shares
LRCX: 2 shares

Total Invested: $42063.35


In [56]:
import yfinance as yf

# Function to calculate weekly return
def calculate_weekly_return(data):
    return (data[-1] - data[0]) / data[0]

# List of top and bottom tickers on 2024-06-21
top_tickers = ['AMT', 'AOS', 'ARE', 'ARGX', 'AWK', 'BEP', 'BIDU', 'FND', 'BIP', 'BLD']
bottom_tickers = ['HUBS', 'CAVA', 'WING', 'NTNX', 'CMG', 'MNDY', 'FTAI', 'META', 'NFLX', 'NOW']

# Fetch data from yfinance for the given date range
start_date = '2024-07-14'
end_date = '2024-07-20'

tickers = top_tickers + bottom_tickers
data = yf.download(tickers, start=start_date, end=end_date)

# Calculate weekly returns
weekly_returns = {}
for ticker in tickers:
    ticker_data = data['Adj Close'][ticker]
    weekly_returns[ticker] = calculate_weekly_return(ticker_data)

# Display the actual weekly returns
print("Actual Weekly Returns from 2024-07-14 to 2024-07-20")

print("\nTop 10 Predicted Tickers on 2024-07-12:")
for ticker in top_tickers:
    print(f"{ticker}: {weekly_returns[ticker]:.6f}")

print("\nBottom 10 Predicted Tickers on 2024-07-12:")
for ticker in bottom_tickers:
    print(f"{ticker}: {weekly_returns[ticker]:.6f}")


[*********************100%%**********************]  20 of 20 completed
Actual Weekly Returns from 2024-07-14 to 2024-07-20

Top 10 Predicted Tickers on 2024-07-12:
AMT: 0.006225
AOS: 0.002164
ARE: -0.014560
ARGX: 0.020306
AWK: 0.035429
BEP: -0.061256
BIDU: -0.029844
FND: 0.023194
BIP: -0.003599
BLD: 0.022260

Bottom 10 Predicted Tickers on 2024-07-12:
HUBS: 0.016805
CAVA: -0.076648
WING: -0.043734
NTNX: -0.066486
CMG: -0.054231
MNDY: -0.033315
FTAI: 0.003638
META: -0.039040
NFLX: -0.035204
NOW: -0.021684


In [57]:
import yfinance as yf

# Function to calculate weekly return
def calculate_weekly_return(data):
    return (data[-1] - data[0]) / data[0]

# Function to calculate the new portfolio value
def calculate_portfolio_value(initial_investment, returns, short=False):
    if short:
        returns = -returns  # Invert the returns for short positions
    return initial_investment * (1 + returns)

# List of top and bottom tickers on 2024-06-21
top_tickers = ['AMT', 'AOS', 'ARE', 'ARGX', 'AWK', 'BEP', 'BIDU', 'FND', 'BIP', 'BLD']
bottom_tickers = ['HUBS', 'CAVA', 'WING', 'NTNX', 'CMG', 'MNDY', 'FTAI', 'META', 'NFLX', 'NOW']

# Fetch data from yfinance for the given date range
start_date = '2024-07-14'
end_date = '2024-07-20'

tickers = top_tickers + bottom_tickers
data = yf.download(tickers, start=start_date, end=end_date)

# Calculate weekly returns
weekly_returns = {}
for ticker in tickers:
    ticker_data = data['Adj Close'][ticker]
    weekly_returns[ticker] = calculate_weekly_return(ticker_data)

# Initialize portfolio value
initial_portfolio_value = 40000  # Total initial portfolio value
individual_investment = initial_portfolio_value / len(tickers)

# Calculate new portfolio value
new_portfolio_value = 0
for ticker in top_tickers:
    new_value = calculate_portfolio_value(individual_investment, weekly_returns[ticker], short=False)
    new_portfolio_value += new_value

for ticker in bottom_tickers:
    new_value = calculate_portfolio_value(individual_investment, weekly_returns[ticker], short=True)
    new_portfolio_value += new_value

# Display the actual weekly returns
print("Actual Weekly Returns from 2024-07-14 to 2024-07-20")

print("\nTop 10 Predicted Tickers on 2024-07-20:")
for ticker in top_tickers:
    print(f"{ticker}: {weekly_returns[ticker]:.6f}")

print("\nBottom 10 Predicted Tickers on 2024-07-20:")
for ticker in bottom_tickers:
    print(f"{ticker}: {weekly_returns[ticker]:.6f}")

# Display the portfolio value
print(f"\nInitial Portfolio Value: ${initial_portfolio_value}")
print(f"New Portfolio Value: ${new_portfolio_value:.2f}")
print(f"Total Return: {((new_portfolio_value - initial_portfolio_value) / initial_portfolio_value) * 100:.2f}%")


[*********************100%%**********************]  20 of 20 completed
Actual Weekly Returns from 2024-07-14 to 2024-07-20

Top 10 Predicted Tickers on 2024-07-20:
AMT: 0.006225
AOS: 0.002164
ARE: -0.014560
ARGX: 0.020306
AWK: 0.035429
BEP: -0.061256
BIDU: -0.029844
FND: 0.023194
BIP: -0.003599
BLD: 0.022260

Bottom 10 Predicted Tickers on 2024-07-20:
HUBS: 0.016805
CAVA: -0.076648
WING: -0.043734
NTNX: -0.066486
CMG: -0.054231
MNDY: -0.033315
FTAI: 0.003638
META: -0.039040
NFLX: -0.035204
NOW: -0.021684

Initial Portfolio Value: $40000
New Portfolio Value: $40700.43
Total Return: 1.75%


This is the code to determine equal shares.

# Below this is a XGBoost Ranker model that has worse returns. .05% weekly return.

In [54]:
# Load the CSV file, preserving the Date and ticker multi-index
all_data = pd.read_csv('mega_historical_data4.csv', index_col=['Date', 'ticker'], parse_dates=['Date'])

In [59]:
import xgboost as xgb
import pandas as pd
import numpy as np

# Step 1: Prepare the Data
# Assuming 'all_data' is your multi-index DataFrame with Date and ticker

# Flatten the multi-index for compatibility with XGBoost
all_data = all_data.reset_index()

# Define features and target
features = ['3Day_Ret_Class', 'Weekly_Ret_Class', 'RSI', 'ADX', 'Weekly_Return_Lag_1', 'Weekly_Return_Lag_2',
            'Weekly_Return_Lag_3', 'Weekly_Return_Lag_4', 'Weekly_Return_Lag_8', 'Weekly_Volatility', 
            'Weekly_RVOL', 'Weekly_ROC', 'MACD_8_24', 'MACD_16_24', 'MACD_16_96', 'Weekly_Return_Lag_1_Rank']  
target = 'Weekly_Return_Rank'  # Relevance is now Weekly_Return_Rank

# Ensure the relevance degrees are within the acceptable range
all_data[target] = all_data[target].clip(upper=31)

# Invert the relevance labels to ensure higher ranks are treated as more relevant
max_rank = all_data[target].max()
all_data[target] = max_rank - all_data[target]

# Split the data into training and testing sets
train_data = all_data[all_data['Date'] < '2022-07-12']
test_data = all_data[all_data['Date'] >= '2022-07-12']

# Define qids (query ids) for the groups based on the Date
train_qids = train_data['Date'].factorize()[0]
test_qids = test_data['Date'].factorize()[0]

# Create group lists for training and testing data
train_group = train_data['Date'].value_counts().sort_index().values
test_group = test_data['Date'].value_counts().sort_index().values

# Define the DMatrix for XGBoost
dtrain = xgb.DMatrix(train_data[features], label=train_data[target])
dtrain.set_group(train_group)
dtest = xgb.DMatrix(test_data[features], label=test_data[target])
dtest.set_group(test_group)

# Step 2: Train the XGBRanker Model
ranker = xgb.XGBRanker(
    objective='rank:pairwise',  # Pairwise ranking objective
    eval_metric='ndcg',         # NDCG metric for evaluation
    eta=0.1,
    gamma=1.0,
    min_child_weight=0.1,
    max_depth=6,
    nthread=4,
    tree_method='hist',
    early_stopping_rounds=10,
    device='cuda'  # Use GPU acceleration if available
)

# Train the model
ranker.fit(train_data[features], train_data[target], group=train_group, eval_set=[(test_data[features], test_data[target])], eval_group=[test_group])

# Step 3: Evaluate and Predict
predicted_scores = ranker.predict(test_data[features])

# Convert predicted scores to ranks
test_data['Predicted_Score'] = predicted_scores
test_data['Predicted_Rank'] = test_data.groupby('Date')['Predicted_Score'].rank(method='first', ascending=False)

# Restore multi-index
test_data = test_data.set_index(['Date', 'ticker'])

# Select a specific date for inspection
specific_date = '2024-07-19'
data_for_date = test_data.xs(specific_date, level='Date')

# Display the top 10 and bottom 10
top_10 = data_for_date.nsmallest(10, 'Predicted_Rank')
bottom_10 = data_for_date.nlargest(10, 'Predicted_Rank')

print("Top 10 Tickers by Predicted Rank on", specific_date)
print(top_10[['Weekly_Return', 'Weekly_Return_Rank', 'Predicted_Score', 'Predicted_Rank']])

print("\nBottom 10 Tickers by Predicted Rank on", specific_date)
print(bottom_10[['Weekly_Return', 'Weekly_Return_Rank', 'Predicted_Score', 'Predicted_Rank']])

XGBoostError: [11:29:50] /workspace/src/common/ranking_utils.h:348: Check failed: label_is_integer: When using relevance degree as target, label must be either 0 or positive integer.
Stack trace:
  [bt] (0) /home/annalisa-aaron/anaconda3/envs/quant_env/lib/python3.8/site-packages/xgboost/lib/libxgboost.so(+0x1ba24e) [0x7aa697bba24e]
  [bt] (1) /home/annalisa-aaron/anaconda3/envs/quant_env/lib/python3.8/site-packages/xgboost/lib/libxgboost.so(+0x81d6dd) [0x7aa69821d6dd]
  [bt] (2) /home/annalisa-aaron/anaconda3/envs/quant_env/lib/python3.8/site-packages/xgboost/lib/libxgboost.so(+0x804fb0) [0x7aa698204fb0]
  [bt] (3) /home/annalisa-aaron/anaconda3/envs/quant_env/lib/python3.8/site-packages/xgboost/lib/libxgboost.so(+0x525225) [0x7aa697f25225]
  [bt] (4) /home/annalisa-aaron/anaconda3/envs/quant_env/lib/python3.8/site-packages/xgboost/lib/libxgboost.so(+0x525484) [0x7aa697f25484]
  [bt] (5) /home/annalisa-aaron/anaconda3/envs/quant_env/lib/python3.8/site-packages/xgboost/lib/libxgboost.so(+0x525c58) [0x7aa697f25c58]
  [bt] (6) /home/annalisa-aaron/anaconda3/envs/quant_env/lib/python3.8/site-packages/xgboost/lib/libxgboost.so(+0x525e43) [0x7aa697f25e43]
  [bt] (7) /home/annalisa-aaron/anaconda3/envs/quant_env/lib/python3.8/site-packages/xgboost/lib/libxgboost.so(+0x4c8ab0) [0x7aa697ec8ab0]
  [bt] (8) /home/annalisa-aaron/anaconda3/envs/quant_env/lib/python3.8/site-packages/xgboost/lib/libxgboost.so(XGBoosterEvalOneIter+0x2f6) [0x7aa697b6abb6]



In [ ]:
# Select a specific date for inspection
specific_date = '2024-07-12'
data_for_date = test_data.xs(specific_date, level='Date')

# Display the top 10 and bottom 10
top_10 = data_for_date.nsmallest(10, 'Predicted_Rank')
bottom_10 = data_for_date.nlargest(10, 'Predicted_Rank')

print("Top 10 Tickers by Predicted Rank on", specific_date)
print(top_10[['Weekly_Return', 'Weekly_Return_Rank', 'Predicted_Score', 'Predicted_Rank']])

print("\nBottom 10 Tickers by Predicted Rank on", specific_date)
print(bottom_10[['Weekly_Return', 'Weekly_Return_Rank', 'Predicted_Score', 'Predicted_Rank']])

In [ ]:
import yfinance as yf

# Function to calculate weekly return
def calculate_weekly_return(data):
    return (data[-1] - data[0]) / data[0]

# Function to calculate the new portfolio value
def calculate_portfolio_value(initial_investment, returns, short=False):
    if short:
        returns = -returns  # Invert the returns for short positions
    return initial_investment * (1 + returns)

# List of top and bottom tickers on 2024-06-21
top_tickers = ['SRPT', 'ADBE', 'SIRI', 'RYAN', 'TOST', 'GILD', 'UTHR', 'ACN', 'CHWY', 'ORLY']
bottom_tickers = ['YUMC', 'UWMC', 'UHAL', 'UAL', 'TM', 'TAK', 'SYM', 'STM', 'STLA', 'SQM']

# Fetch data from yfinance for the given date range
start_date = '2024-06-23'
end_date = '2024-06-29'

tickers = top_tickers + bottom_tickers
data = yf.download(tickers, start=start_date, end=end_date)

# Calculate weekly returns
weekly_returns = {}
for ticker in tickers:
    ticker_data = data['Adj Close'][ticker]
    weekly_returns[ticker] = calculate_weekly_return(ticker_data)

# Initialize portfolio value
initial_portfolio_value = 40000  # Total initial portfolio value
individual_investment = initial_portfolio_value / len(tickers)

# Calculate new portfolio value
new_portfolio_value = 0
for ticker in top_tickers:
    new_value = calculate_portfolio_value(individual_investment, weekly_returns[ticker], short=False)
    new_portfolio_value += new_value

for ticker in bottom_tickers:
    new_value = calculate_portfolio_value(individual_investment, weekly_returns[ticker], short=True)
    new_portfolio_value += new_value

# Display the actual weekly returns
print("Actual Weekly Returns from 2024-07-07 to 2024-07-12")

print("\nTop 10 Predicted Tickers on 2024-07-12:")
for ticker in top_tickers:
    print(f"{ticker}: {weekly_returns[ticker]:.6f}")

print("\nBottom 10 Predicted Tickers on 2024-07-12:")
for ticker in bottom_tickers:
    print(f"{ticker}: {weekly_returns[ticker]:.6f}")

# Display the portfolio value
print(f"\nInitial Portfolio Value: ${initial_portfolio_value}")
print(f"New Portfolio Value: ${new_portfolio_value:.2f}")
print(f"Total Return: {((new_portfolio_value - initial_portfolio_value) / initial_portfolio_value) * 100:.2f}%")



Below is code with similar features (RSI, ADX, etc.) that gave better returns. Trying to replicate (except this has weekly returns and that had 126 day).

The differences are a bigger stock universe (2625) and these features: return for the last 6 months (pct change 126), return lag 1-4, 14 day RSI, 14 day ADX, 20 day RVOL, Analyst Rating, Quarterly Revenue Growth YoY, Distance from 52 week High, Distance from 52 week Low. Target as Return. 